enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+
+### Goals
+
+The goal is to provide OpenShift users/admins an easy way to run and trigger Insights data gathering on demand and refresh the Insights status of their cluster/s. 
+The way of triggering the new data gathering should be simple enough to e.g create a new button in the ACM (advanced cluster manager) refreshing the Insights status of the given cluster. 
+
+### Non-Goals
+
+- Change data gather frequency.
+- Change gathered data. 
+- Change operator status reporting (including metrics). This should work as previously.
+
+## Proposal
+
+The enhancement proposal introduces a new Insights `CustomResourceDefinition` type in the OpenShift API that will serve as the definition and trigger to new data gathering on demand. 
+This new CRD will build on and extend the existing `GatherConfig` (see the [Insights config API enhancement](../insights/insights-config-api.md)). The idea is that with each new CR instance of this resource, new data gathering will happen. 
+
+The proposal for the new `datagather.insights.openshift.io` resource is:
+
+```go
+type DataGather struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	Spec DataGatherSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status DataGatherStatus `json:"status"`
+}
+
+type DataGatherSpec struct {
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "None" and "ObfuscateNetworking".
+	// When set to None the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is None.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy"`
+	// disabledGatherers is a list of gatherers to be excluded from the gathering. All the gatherers can be disabled by providing "all" value.
+	// If all the gatherers are disabled, the Insights operator does not gather any data.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
+	// +optional
+	DisabledGatherers []string `json:"disabledGatherers"`
```

Comment:
I suspect you'll eventually want configuration per gatherer. Such configuration doesn't have to be added in this PR, but I'd like to know how you'd structure it a year from now before merging the API PR.

Reply:
Interesting point. I can't envision such configuration right now, but I guess the suggestion is to do something like:
```
type Gatherer struct {
 name string
}
```
and use it instead of `string`, right?

Reply:
Maybe instead you'd want
```
type GathererConfig struct {
  Name string
  State: Enabled | Disabled
  ... <other options later>
}
```
That would allow a generic configuration, but if gatherers are likely to need individual configuration at some point then we would want to include a discriminated union in there too
```
type GathererConfig struct {
  Name string (Enum?)
  State Enum (Enabled | Disabled)
  Config {
    Name Enum (Foo | Bar| Baz)
    Foo *FooConfig
    Bar *BarConfig
    Baz *BazConfig
  }
}
```

It would make the enabled/disabled decision a property of the gatherer's config rather than being separate

Reply:
OK thanks. I was thinking about it and the main reason for introducing this attribute was to allow the data collection to be turned off completely (in the first config API see [here](https://github.com/openshift/api/blob/master/config/v1alpha1/types_insights.go#L45)) so I think I would prefer to keep it as it is for now.

Reply:
Revising the structure now, and having something like I've suggested above, but keeping it simple, so for now it would be
```
gatherers:
- name: MyGatherA
  state: Disabled
```
would allow you to evolve in the future, where keeping the disabled list might mean the configuration of this in the future is split between `disabledGatherers` and `gatherers` lists

Reply:
This all makes sense to me, but I am failing to see a simple way to disable all the gatherers (with this approach). This is now easily doable with `disabledGatherers: all` and this was the main requirement.

Reply:
That's a fair point, you wouldn't be able to do that with the combined configuration option, which is a clear disadvantage of that approach

@deads2k any ideas on that discrepancy?

Reply:
I've been thinking about it again and maybe we could really go with this suggestion, because my question was why someone would force a gathering with all gatherers disabled? It doesn't make much sense right? 

This option (to disable gathering completely) is a must for the periodical one (for the https://github.com/openshift/api/blob/master/config/v1alpha1/types_insights.go#L52), but here I think it doesn't make much sense....

Reply:
Yeah, I'm wondering now if we want to make a generic `gatherState: Enabled|Disabled|Cron` kind of enum, and then have the config for individual gatherers as you've suggested here

The other API is still alpha right? We could make that change as we promote it

Reply:
Yes, still alpha. I am open to any reasonable changes, but I just want to have a very simple way to turn off/disable the gathering completely.

Reply:
I created https://issues.redhat.com/browse/CCXDEV-10455 to track this